File: bin/rice_cooker.dart
import 'dart:io';

enum RiceCookerState { idle, cooking, warm }

enum Commands { cook, warm, cancel, addWatter, plugIn, unPlug, exit }

enum ErrorCodes { noWatter, notPluggedIn, busy }

class RiceCooker {
  RiceCookerState state = RiceCookerState.idle;
  bool hasWater = true;
  bool isPluggedIn = true;

  Future<void> cookRice() async {
    if (!isPluggedIn) {
      print('Error: ${ErrorCodes.notPluggedIn} - Rice cooker is not plugged in.');
      return;
    }

    if (!hasWater) {
      print('Error: ${ErrorCodes.noWatter} - No water detected. Please add water and try again.');
      return;
    }

    if (state != RiceCookerState.idle) {
      print('Rice cooker is already busy.');
      return;
    }

    state = RiceCookerState.cooking;
    print('Cooking rice...');

    // Simulate cooking time
    await Future.delayed(Duration(seconds: 3));

    print('Rice is cooked!');
    state = RiceCookerState.idle;
  }

  Future<void> warmRice() async {
    if (!isPluggedIn) {
      print('Error: ${ErrorCodes.notPluggedIn} - Rice cooker is not plugged in.');
      return;
    }

    if (!hasWater) {
      print('Error: ${ErrorCodes.noWatter} - No water detected. Please add water and try again.');
      return;
    }

    if (state != RiceCookerState.idle && state != RiceCookerState.warm) {
      print('Rice cooker is not ready for warming.');
      return;
    }

    state = RiceCookerState.warm;
    print('Warming rice...');

    // Simulate warming time
    await Future.delayed(Duration(seconds: 1));

    print('Rice is warm and ready to serve!');
    state = RiceCookerState.idle;
  }

  void cancel() {
    if (state == RiceCookerState.idle) {
      print('Nothing to cancel.');
      return;
    }

    print('Canceling operation...');
    state = RiceCookerState.idle;
    print('Operation canceled.');
  }

  void addWater() {
    hasWater = true;
    print('Water added successfully.');
  }

  void plugIn() {
    isPluggedIn = true;
    print('Rice cooker plugged in.');
  }

  void unplug() {
    if (state != RiceCookerState.idle) {
      print('Error: ${ErrorCodes.busy} - Rice cooker is busy. Please wait until it finishes before unplugging.');
      return;
    }

    isPluggedIn = false;
    print('Rice cooker unplugged.');
  }

  void printMenu() {
    print('\nMenu:');
    print('1. Cook Rice');
    print('2. Warm Rice');
    print('3. Cancel');
    print('4. Add Water');
    print('5. Plug In');
    print('6. Unplug');
    print('7. Exit');
  }

  Future<void> handleCommand(int command) async {
    switch (command) {
      case 1:
        await cookRice();
        break;
      case 2:
        await warmRice();
        break;
      case 3:
        cancel();
        break;
      case 4:
        addWater();
        break;
      case 5:
        plugIn();
        break;
      case 6:
        unplug();
        break;
      case 7:
        print('Exiting Rice Cooker CLI. Goodbye!');
        exit(0);
      default:
        print('Invalid command. Please enter a valid option.');
    }
  }
}

void main() async {
  final cooker = RiceCooker();

  while (true) {
    cooker.printMenu();
    stdout.write('Enter your command: ');
    final userChoice = stdin.readLineSync();

    if (userChoice == null) {
      print('Invalid input. Please enter a command.');
      continue;
    }

    final command = int.tryParse(userChoice.trim());

    if (command == null) {
      print('Invalid input. Please enter a number.');
    } else {
      await cooker.handleCommand(command);
    }
  }
}

File: RiceCooker.ts
import { RiceCookerState, ErrorCodes, Commands } from './RiceCookerState'

export class RiceCooker {
  private state: RiceCookerState = RiceCookerState.IDLE
  private hasWater: boolean = false
  private isPluggedIn: boolean = false

  cookRice (): void {
    if (!this.isPluggedIn) {
      console.error(
        `Error: ${ErrorCodes.NOT_PLUGGED_IN} - Rice cooker is not plugged in.`
      )
    } else if (!this.hasWater) {
      console.error(
        `Error: ${ErrorCodes.NO_WATER} - No water detected. Please add water and try again.`
      )
    } else if (this.state !== RiceCookerState.IDLE) {
      console.log('Rice cooker is already busy.')
    } else {
      this.state = RiceCookerState.COOKING
      console.log('Cooking rice...')

      this.cookRiceAsync()
        .then((value) => {
          console.log('Rice is cooked!')
        })
        .catch((error) => {
          console.error(`Error: ${error as string}`)
        })
        .finally(() => {
          this.state = RiceCookerState.IDLE
        })
    }
  }

  private async cookRiceAsync (): Promise<void> {
    return await new Promise((resolve) => {
      setTimeout(() => {
        resolve()
      }, 3000)
    })
  }

  warmRice (): void {
    if (!this.isPluggedIn) {
      console.error(
        `Error: ${ErrorCodes.NOT_PLUGGED_IN} - Rice cooker is not plugged in.`
      )
    } else if (!this.hasWater) {
      console.error(
        `Error: ${ErrorCodes.NO_WATER} - No water detected. Please add water and try again.`
      )
    } else if (
      this.state !== RiceCookerState.IDLE &&
      this.state !== RiceCookerState.WARM
    ) {
      console.log('Rice cooker is not ready for warming.')
    } else {
      this.state = RiceCookerState.WARM
      console.log('Warming rice...')

      // Simulate warming time
      setTimeout(() => {
        console.log('Rice is warm and ready to serve!')
        this.state = RiceCookerState.IDLE
      }, 1000)
    }
  }

  cancel (): void {
    if (this.state === RiceCookerState.IDLE) {
      console.log('Nothing to cancel.')
    } else {
      console.log('Canceling operation...')
      this.state = RiceCookerState.IDLE
      console.log('Operation canceled.')
    }
  }

  addWater (): void {
    this.hasWater = true
    console.log('Water added successfully.')
  }

  plugIn (): void {
    this.isPluggedIn = true
    console.log('Rice cooker plugged in.')
  }

  unplug (): void {
    if (this.state !== RiceCookerState.IDLE) {
      console.error(
        `Error: ${ErrorCodes.BUSY} - Rice cooker is busy. Please wait until it finishes before unplugging.`
      )
    } else {
      this.isPluggedIn = false
      console.log('Rice cooker unplugged.')
    }
  }

  printMenu (): void {
    console.log('\nMenu:')
    console.log('1. Cook Rice')
    console.log('2. Warm Rice')
    console.log('3. Cancel')
    console.log('4. Add Water')
    console.log('5. Plug In')
    console.log('6. Unplug')
  }

  handleCommand (command: Commands): void {
    switch (command) {
      case Commands.COOK:
        this.cookRice()
        break
      case Commands.WARM:
        this.warmRice()
        break
      case Commands.CANCEL:
        this.cancel()
        break
      case Commands.ADD_WATER:
        this.addWater()
        break
      case Commands.PLUG_IN:
        this.plugIn()
        break
      case Commands.UNPLUG:
        this.unplug()
        break
      default:
        console.log('Invalid command. Please enter a valid option.')
    }
  }
}
